README.md
![Build Status](https://github.com/iquerejeta/Poseidon252/workflows/Continuous%20integration/badge.svg)
[![Repository](https://img.shields.io/badge/github-poseidon252-blueviolet)](https://github.com/iquerejeta/Poseidon252)

# Dusk-Poseidon
Fork from Dusk Network's [original](https://github.com/dusk-network/Poseidon252) implementation. 
Reference implementation for the Poseidon Hashing algorithm.

#### Reference

[Starkad and Poseidon: New Hash Functions for Zero Knowledge Proof Systems](https://eprint.iacr.org/2019/458.pdf)

This repository has been created so there's a unique library that holds the tools & functions
required to perform Poseidon Hashes.

This hashes heavily rely on the Hades permutation, which is one of the key parts that Poseidon needs in order
to work.
This library uses the reference implementation of [Dusk-Hades](https://github.com/dusk-network/hades252) which has been
designed & build by the [Dusk-Network team](https://dusk.network/).

**The library provides the two hashing techniques of Poseidon:**

## Sponge Hash

The `Sponge` technique in Poseidon allows to hash an unlimited amount of data
into a single `Scalar`.
The sponge hash technique requires a padding to be applied before the data can
be hashed.

This is done to avoid hash collisions as stated in the paper of the Poseidon Hash
algorithm. See: <https://eprint.iacr.org/2019/458.pdf>.
The inputs of the `sponge_hash` are always `Scalar` or need to be capable of being represented
as it.

The module provides two sponge hash implementations:

- Sponge hash using `Scalar` as backend. Which hashes the inputted `Scalar`s and returns a single
  `Scalar`.

- Sponge hash gadget using `plonk::Witness` as a backend. This technique is used/required
  when you want to proof pre-images of unconstrained data inside Zero-Knowledge PLONK circuits.

## Merkle Hash

The Merkle Level Hashing is a technique that Poseidon is optimized-by-design
to perform.
This technique allows us to perform hashes of an entire Merkle Tree using
`Dusk-Hades` as backend.

The technique requires the computation of a `bitflags` element which is always
positioned as the first item of the level when we hash it, and it basically generated
in respect of the presence or absence of a leaf in the tree level.
This allows to prevent hashing collisions.

At the moment, this library is designed and optimized to work only with trees of `ARITY`
up to 4. **That means that trees with a bigger ARITY SHOULD NEVER be used with this lib.**
The module contains the implementation of 4 variants of the same algorithm to support the
majority of the configurations that the user may need:

- Scalar backend for hashing Merkle Tree levels outside ZK-Circuits with two variants:
  One of them computes the bitflags item while the other assumes that it has already been
  computed and placed in the first Level position.

- `plonk::Witness` backend for hashing Merkle Tree levels inside ZK-Circuits,
  specifically, PLONK circuits. This implementation comes also with two variants;
  One of them computes the bitflags item while the other assumes that it has already been
  computed and placed in the first Level position.

### Zero Knowledge Merkle Opening Proof example:

```rust
use plonk::error::Error as PlonkError;
use poseidon::tree::{self, PoseidonBranch, PoseidonLeaf, PoseidonTree};
use rand::rngs::OsRng;
use rand::{CryptoRng, RngCore};


use plonk::prelude::*;
use nstack::annotation::Keyed;

// Depth of the merkle tree
const DEPTH: usize = 17;

// Capacity of the circuit
const CAPACITY: usize = 15;

// Alias for the default tree implementation
type Tree = PoseidonTree<DataLeaf, (), DEPTH>;

// Leaf representation
#[derive(Debug, Default, Clone, Copy, PartialOrd, Ord, PartialEq, Eq)]
pub struct DataLeaf {
    data: BlsScalar,
    pos: u64,
}

// Keyed needs to be implemented for a leaf type and the tree key.
impl Keyed<()> for DataLeaf {
    fn key(&self) -> &() {
        &()
    }
}

impl DataLeaf {
    pub fn random<R: RngCore + CryptoRng>(rng: &mut R) -> Self {
        let data = BlsScalar::random(rng);
        let pos = 0;

        Self { data, pos }
    }
}

// Any leaf of the poseidon tree must implement `PoseidonLeaf`
impl PoseidonLeaf for DataLeaf {
    // Cryptographic hash of the data leaf
    fn poseidon_hash(&self) -> BlsScalar {
        self.data
    }

    // Position on the tree
    fn pos(&self) -> &u64 {
        &self.pos
    }

    // Method used to set the position on the tree after the
    // `PoseidonTree::push` call.
    fn set_pos(&mut self, pos: u64) {
        self.pos = pos;
    }
}

#[derive(Default)]
struct MerkleOpeningCircuit {
    branch: PoseidonBranch<DEPTH>,
}

impl MerkleOpeningCircuit {
    /// Generate N random leaves and append them to the tree
    pub fn random<R: RngCore + CryptoRng>(
        rng: &mut R,
        tree: &mut Tree,
    ) -> Self {
        const N: u64 = 1024;

        // Append 1024 elements to the tree
        for _ in 0..N {
            let leaf = DataLeaf::random(rng);
            tree.push(leaf);
        }

        let branch = tree.branch(N - 1).expect(
            "Failed to fetch the branch of the created leaf from the tree",
        );

        Self { branch }
    }
}

impl Circuit for MerkleOpeningCircuit {
    fn circuit<C>(&self, composer: &mut C) -> Result<(), PlonkError> 
    where 
        C: Composer,
    {
      let leaf: BlsScalar = *self.branch;
      let leaf = composer.append_witness(leaf);
  
      let root = self.branch.root();
      let root = composer.append_witness(*root);
  
      let root_p =
              tree::merkle_opening::<C, DEPTH>(composer, &self.branch, leaf);
  
      composer.assert_equal(root_p, root);
  
      Ok(())
    }
}

// Create a prover and a verifier
let label = b"dusk-network";
let pp = PublicParameters::setup(1 << CAPACITY, &mut OsRng).unwrap();

let (prover, verifier) =
    Compiler::compile(&pp, label).expect("failed to compile circuit");

// Instantiate a new tree
let mut tree = Tree::default();
let circuit = MerkleOpeningCircuit::random(&mut OsRng, &mut tree);

// Generate a ZK opening proof
let (proof, public_inputs) = prover.prove(&mut OsRng, &circuit)
            .expect("proving the circuit should succeed");

// Verify the proof
verifier.verify(&proof, &public_inputs)
    .expect("verifying the proof should succeed");
```

## Documentation

This crate contains info about all the functions that the library provides as well as the
documentation regarding the data structures that it exports. To check it, please feel free to go to
the [documentation page](https://dusk-network.github.io/Poseidon252/poseidon252/index.html)

## Licensing

This code is licensed under Mozilla Public License Version 2.0 (MPL-2.0). Please see [LICENSE](https://github.com/dusk-network/plonk/blob/master/LICENSE) for further info.

## About

Implementation designed by the [dusk](https://dusk.network) team.

## Contributing

- If you want to contribute to this repository/project please, check [CONTRIBUTING.md](https://github.com/dusk-network/Poseidon252/blob/master/CONTRIBUTING.md)
- If you want to report a bug or request a new feature addition, please open an issue on this repository.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [0.28.1] - 2023-01-18

### Added

- Implement `bytes::Serializable` for `PoseidonBranch` and `PoseidonLevel` [#203]
- Add benchmarks for merkle opening proof [#197]

### Changed

- Derive `Copy` for `PoseidonBranch` [#200]

## [0.28.0] - 2022-11-10

### Changed

- Update `dusk-plonk` from `0.12` to `0.13`
- Update `dusk-hades` from `0.19` to `0.20`

## [0.27.0] - 2022-10-19

### Added

- Add support for `rkyv-impl` under `no_std`
- Add `ranno` version `0.1` to dependencies [#180]

### Changed

- Change `PoseidonBranch` to have two fields - `root` and `path`. The path is
  now a fixed length array. [#189]
- Change `PoseidonTree` to build only with the `alloc` feature [#180]
- Change `PoseidonTree` to take a generic `Keyed` over the leaf type
  instead of a `PoseidonAnnotation` [#180]
- Make `PoseidonTree::new` const [#180]
- Update `microkelvin` from `0.15` to `0.17` [#180]
- Update `nstack` from `0.14.0-rc` to `0.16` [#180]

### Removed

- Remove `PoseidonBranch` `Default` implementation [#189]
- Remove `std` feature [#180]
- Remove `canon` and `persistence` features [#180]
- Remove `Error` struct [#180]
- Remove `canonical` and `canonical-derive` from dependencies [#180]
- Remove `PoseidonMaxAnnotation` [#180]

### Fixed

- Fix merkle opening circuit [#181]
- Fix CHANGELOG version links [#191]

## [0.26.0] - 2022-08-17

### Added

- Add `rkyv` implementation behind feature gate [#175]

### Changed

- Update `dusk-bls12_381` from `0.8` to `0.11`
- Update `dusk-jubjub` from `0.10` to `0.12`
- Update `dusk-hades` from `0.17.0-rc` to `0.19`
- Update `canonical` from `0.6` to `0.7`
- Update `canonical_derive` from `0.6` to `0.7`
- Update `microkelvin` from `0.14` to `0.15`
- Update `nstack` from `0.13` to `0.14.0-rc`
- Update `dusk-plonk` from `0.9` to `0.12`
- Change merkle opening to constrain leaf [#162]
- Export `sponge::truncated::hash` regardless of `alloc` feature [#167]
- Remove useless `let` in `sponge::truncated`

### Fixed

- Fix module injection for `tree` and `cipher` modules

## [0.22.0] - 2021-07-27

### Changed

- Update `microkelvin` from `0.6` to `0.10` [#158]
- Update `nstack` from `0.9` to `0.10` [#158]

## [0.21.0] - 2021-07-05

### Added

- Add integration tests with examples of custom walker iterators [#134]
- Add `persistance` feature to the crate [#151]
- Add `truncated` module in sponge to deal with scalar conversions [#153]

### Changed

- Change the tree logic to be compatible with `microkelvin v0.9` [#151]
- Changed toolchain-file version to nightly-2021-06-06 [#149]
- Change feature set config for the crate [#138]
- Update `error` module to be no_std compatible [#132]
- Update to latest `dusk-poseidon`, `dusk-bls12_381` and `dusk-jubjub` [#126]
- Update to latest `microkelvin v0.9`, `nstack v0.9` and `canonical v0.6` [#125]
- Update randomness provider to `rand_core` [#127]
- Change trait bound system for `PoseidonTree` [#125]
- Update `PoseidonTreeAnnotation` to be an autotrait [#125]
- Update feature system for the crate [#138]
- Change `PoseidonLeaf` getter methods to return refs [#143]

### Removed

- Remove `anyhow` and `thiserror` from deps [#132]
- Remove `PoseidonWalkableIterator` and `PoseidonWalkableAnnotation` [#125]
- Remove `canon_host` feature checks from CI [#136]
- Remove `anyhow` and `thiserror` usage [#132]
- Remove `microkelvin` requirements from Tree [#146]

### Fixed

- Fix Readme.md import from lib.rs [#148]

## [0.20.0] - 2021-04-06

### Changed

- Update `dusk-plonk` from `0.6` to `0.7` [#119]
- Update `dusk-hades` from `0.14` to `0.15` [#119]

### Fixed

- Merkle Opening constant circuit description [#122]

## [0.19.0] - 2021-03-11

### Changed

- Update `dusk-plonk` from `0.5` to `0.6` [#117]
- Update `dusk-hades` from `0.13` to `0.14`

## [0.18.0] - 2021-02-11

### Changed

- Change crate's name from `Poseidon252` to `dusk-poseidon`
- Implement `Canon` for `PoseidonBranch`

## [0.17.0] - 2021-02-01

### Changed

- Sponge gadget import path simplified.

### Fixed

- PoseidonBranch minimum depth fixed. [#112]

## [0.16.0] - 2021-01-27

### Changed

- Update canonical to v0.5
- Update dusk-bls12_381 to v0.6
- Update dusk-jubjub to v0.8
- Update dusk-plonk to v0.5
- Update hades252 to v0.12.0
- Remove `hex` crate for unit test in favor of `dusk-bytes`
- Update CHANGELOG to ISO 8601
- to/from bytes methods of `PoseidonCipher` refactored in favor of dusk-bytes

## [0.15.0] - 2020-12-04

### Fixed

- PoseidonMaxAnnotation borrow fix for latest microkelvin.

## [0.14.1] - 2020-11-21

### Changed

- Sponge hash defined as a no-std function.

## [0.14.0] - 2020-11-17

### Changed

- No-Std compatibility.

## [0.13.2] - 2020-11-11

### Changed

- Update Hades version and libraries dependent on BLS and JubJub

## [0.13.1] - 2020-11-06

### Changed

- Feature split between `canon` and `canon_host` for constrained and unconstrained environments.
- Canon implementation for PoseidonTree.

## [0.13.0] - 2020-11-04

### Changed

- PoseidonLeaf pos setter for API consistency with Phoenix

### Removed

- PoseidonCipher std::io implementations

## [0.12.0] - 2020-11-04

### Added

- Gate-featured `canonical` impl.
- `PoseidonAnnotation` as generic to support walkable implementations

## [0.11.0] - 2020-10-30

### Changed

- Bump `hades252` to `v0.10.0`
- Major refactor on the poseidon tree to comply with the simplifications provided by microkelvin

## [0.10.0] - 2020-10-05

### Changed

- Bump `hades252` to `v0.9.0`
- Bump `dusk-plonk` to `v0.3.1`

## [0.9.0] - 2020-10-04

### Added

- `root()` fn for `PoseidonBranch`.

### Changed

- Padding implementation for `PoseidonBranch` with opening gadgets.

### Removed

- Extension fn's from the crate.

## [0.8.1] - 2020-10-01

### Changed


src/tree/branch.rs
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at http://mozilla.org/MPL/2.0/.
//
// Copyright (c) DUSK NETWORK. All rights reserved.

use super::PoseidonLeaf;

use crate::tree::PoseidonAnnotation;

use core::borrow::Borrow;
use core::ops::Deref;

use bls12_381::{Scalar as BlsScalar};
use bytes::{DeserializableSlice, Serializable};
use hades::{ScalarStrategy, Strategy};
use microkelvin::Branch;
use nstack::annotation::Keyed;
use nstack::NStack;

#[cfg(feature = "rkyv-impl")]
use bytecheck::CheckBytes;
#[cfg(feature = "rkyv-impl")]
use rkyv::{Archive, Deserialize, Serialize};

/// Represents a level of a branch on a given depth
#[derive(Debug, Default, Clone, Copy)]
#[cfg_attr(
    feature = "rkyv-impl",
    derive(Archive, Deserialize, Serialize),
    archive_attr(derive(CheckBytes))
)]
pub struct PoseidonLevel {
    pub(crate) level: [BlsScalar; hades::WIDTH],
    index: u64,
}

impl PoseidonLevel {
    /// Represents the offset of a node for a given path produced by a branch
    /// in a merkle opening
    ///
    /// The first position in a level set is represented as offset `1` because
    /// internally the hades permutation prepend the bitflags for merkle opening
    /// consistency
    pub const fn index(&self) -> u64 {
        self.index
    }

    /// Represents the current level offset as a bitflag
    ///
    /// The LSB (least significant bit) represents the offset `1`. Any increment
    /// on the offset will left shift this flag by `1`.
    pub const fn offset_flag(&self) -> u64 {
        1 << (self.index - 1)
    }
}

impl Serializable<{ BlsScalar::SIZE * hades::WIDTH + u64::SIZE }>
    for PoseidonLevel
{
    type Error = bytes::Error;

    fn from_bytes(buf: &[u8; Self::SIZE]) -> Result<Self, Self::Error>
    where
        Self: Sized,
    {
        let mut bytes = &buf[..];

        let mut level = [BlsScalar::zero(); hades::WIDTH];
        for scalar in level.iter_mut() {
            *scalar = BlsScalar::from_reader(&mut bytes)?;
        }

        let index = u64::from_reader(&mut bytes)?;

        Ok(Self { level, index })
    }

    fn to_bytes(&self) -> [u8; Self::SIZE] {
        let mut buf = [0u8; Self::SIZE];

        for (i, scalar) in self.level.iter().enumerate() {
            buf[i * BlsScalar::SIZE..(i + 1) * BlsScalar::SIZE]
                .copy_from_slice(&scalar.to_bytes());
        }

        buf[BlsScalar::SIZE * hades::WIDTH..]
            .copy_from_slice(&self.index.to_bytes());

        buf
    }
}

impl Deref for PoseidonLevel {
    type Target = BlsScalar;

    fn deref(&self) -> &Self::Target {
        &self.level[self.index as usize]
    }
}

impl AsRef<[BlsScalar]> for PoseidonLevel {
    fn as_ref(&self) -> &[BlsScalar] {
        &self.level
    }
}

/// Represents a full path for a merkle opening
#[derive(Debug, Clone, Copy)]
#[cfg_attr(
    feature = "rkyv-impl",
    derive(Archive, Deserialize, Serialize),
    archive_attr(derive(CheckBytes))
)]
pub struct PoseidonBranch<const DEPTH: usize> {
    pub(crate) path: [PoseidonLevel; DEPTH],
    root: BlsScalar,
}

// This macro is necessary due to the fact that `generic_const_exprs`, is still
// unstable. It should remain here until this is completed:
// https://github.com/rust-lang/rust/issues/76560
macro_rules! serializable_branch {
    ($($depth:literal),+) => {
        $(impl Serializable<{ PoseidonLevel::SIZE * $depth + BlsScalar::SIZE }>
            for PoseidonBranch<$depth>
        {
            type Error = bytes::Error;

            fn from_bytes(buf: &[u8; Self::SIZE]) -> Result<Self, Self::Error>
            where
                Self: Sized,
            {
                let mut bytes = &buf[..];

                let mut path = [PoseidonLevel::default(); $depth];
                for level in path.iter_mut() {
                    *level = PoseidonLevel::from_reader(&mut bytes)?;
                }

                let root = BlsScalar::from_reader(&mut bytes)?;

                Ok(Self { path, root })
            }

            fn to_bytes(&self) -> [u8; Self::SIZE] {
                let mut buf = [0u8; Self::SIZE];

                for (i, level) in self.path.iter().enumerate() {
                    buf[i * PoseidonLevel::SIZE..(i + 1) * PoseidonLevel::SIZE]
                        .copy_from_slice(&level.to_bytes());
                }

                buf[PoseidonLevel::SIZE * $depth..]
                    .copy_from_slice(&self.root.to_bytes());

                buf
            }
        })*
    };
}

serializable_branch!(
    1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21,
    22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 64, 128, 256, 512, 1024, 2048
);

impl<const DEPTH: usize> PoseidonBranch<DEPTH> {
    /// Root representation when the tree is empty
    pub const NULL_ROOT: BlsScalar = BlsScalar::zero();

    /// Represents the root for a given path of an opening over a subtree
    pub fn root(&self) -> &BlsScalar {
        &self.root
    }
}

impl<const DEPTH: usize> Default for PoseidonBranch<DEPTH> {
    fn default() -> Self {
        Self {
            path: [PoseidonLevel::default(); DEPTH],
            root: BlsScalar::default(),
        }
    }
}

impl<const DEPTH: usize> Deref for PoseidonBranch<DEPTH> {
    type Target = BlsScalar;

    fn deref(&self) -> &Self::Target {
        self.path[0].deref()
    }
}

impl<const DEPTH: usize> AsRef<[PoseidonLevel]> for PoseidonBranch<DEPTH> {
    fn as_ref(&self) -> &[PoseidonLevel] {
        &self.path
    }
}

impl<L, K, const DEPTH: usize>
    From<&Branch<'_, NStack<L, PoseidonAnnotation<K>>, PoseidonAnnotation<K>>>
    for PoseidonBranch<DEPTH>
where
    L: PoseidonLeaf + Keyed<K>,
    K: Clone + PartialOrd,
{
    fn from(
        b: &Branch<'_, NStack<L, PoseidonAnnotation<K>>, PoseidonAnnotation<K>>,
    ) -> Self {
        let mut path = [PoseidonLevel::default(); DEPTH];

        b.levels().iter().rev().zip(path.iter_mut()).for_each(
            |(nstack_level, poseidon_level)| {
                poseidon_level.index = nstack_level.index() as u64 + 1;

                let mut flag = 1;
                let mut mask = 0;

                match &**nstack_level {
                    NStack::Leaf(l) => l
                        .iter()
                        .zip(poseidon_level.level.iter_mut().skip(1))
                        .for_each(|(leaf, l)| {
                            if let Some(leaf) = leaf {
                                mask |= flag;
                                *l = leaf.poseidon_hash();
                            }

                            flag <<= 1;
                        }),
                    NStack::Node(n) => n
                        .iter()
                        .zip(poseidon_level.level.iter_mut().skip(1))
                        .for_each(|(node, l)| {
                            if let Some(annotated) = node {
                                let anno = annotated.anno();
                                let anno = &*anno;

                                mask |= flag;

                                *l = *anno.borrow();
                            }

                            flag <<= 1;
                        }),
                }

                poseidon_level.level[0] = BlsScalar::from(mask);
            },
        );

        // If the nstack is smaller than the poseidon tree then the we need to
        // populate the remaining levels of the tree.
        let nstack_depth = b.levels().len();
        let flag = BlsScalar::one();

        if nstack_depth < DEPTH {
            let level = path[nstack_depth - 1].level;
            let mut perm = [BlsScalar::zero(); hades::WIDTH];

            let mut h = ScalarStrategy::new();

            path.iter_mut().skip(nstack_depth).fold(level, |l, b| {
                perm.copy_from_slice(&l);
                h.perm(&mut perm);

                b.index = 1;
                b.level[0] = flag;
                b.level[1] = perm[1];

                b.level
            });
        }

        // TODO: The amount of repetition here hints at the fact that hashing
        //  should be more ergonomic.

        // Calculate the root
        let mut perm = [BlsScalar::zero(); hades::WIDTH];
        let mut h = ScalarStrategy::new();

        perm.copy_from_slice(&path[DEPTH - 1].level);
        perm[0] = flag;
        h.perm(&mut perm);

        PoseidonBranch {
            path,
            root: perm[1],
        }
    }
}

#[cfg(test)]
mod tests {
    use crate::tree::{PoseidonBranch, PoseidonLevel};

    use bls12_381::BlsScalar;
    use bytes::Serializable;

    #[test]
    fn branch_serde() {
        type Branch = PoseidonBranch<17>;

        let mut poseidon_level = PoseidonLevel {
            index: 0xbeef,
            ..Default::default()
        };
        poseidon_level.level[3] = BlsScalar::from(42);

        let mut branch = Branch {
            root: BlsScalar::from(4),
            ..Default::default()
        };
        branch.path[7] = poseidon_level;

        let branch_bytes = branch.to_bytes();
        let deserialized_branch = Branch::from_bytes(&branch_bytes)
            .expect("Deserializing should succeed");

        assert_eq!(branch.root, deserialized_branch.root);
        for (level, deserialized_level) in
            branch.path.iter().zip(&deserialized_branch.path)
        {
            assert_eq!(level.index, deserialized_level.index);
            for (scalar, deserialized_scalar) in
                level.level.iter().zip(&deserialized_level.level)
            {
                assert_eq!(scalar, deserialized_scalar);
            }
        }
    }
}
